wallet: let Listwalletdir do not iterate through our blocksdata.

Saibato opened the pull request:
Fix: While fiddling with wallets, I noticed that if datadir == walltetsdir an unsynced node can become quite unresponsive.

if no "wallets" dir is in datadir ( i.e. exiting old node ) 
we would have iterate trough our own node files
to find wallets, that consumes time and could cause an unresponsive node.

Saibato committed:
wallet: let ListWalletDir do not iterate trough our blocksdata.

When WalletDir == DataDir we would have iterate trough our own node files
to find wallets, that consumes time and could cause an unresponsive node.

promag commented:
No strong opinion regarding excluding some paths. Current patch needs to be updated (format) and maybe change implementation to use a list of ignore/exclude paths.

This can improve GUI by reducing freezes but it's not the right fix for that. When `ListWalletDir` is triggered from the GUI, it should be called on a background thread. Alternatively we could consider caching `ListWalletDir` result and invalidate it when wallets are created or loaded.

Saibato commented:
> No strong opinion regarding excluding some paths. Current patch needs to be updated (format) and maybe change implementation to use a list of ignore/exclude paths.
> 
> This can improve GUI by reducing freezes but it's not the right fix for that. When `ListWalletDir` is triggered from the GUI, it should be called on a background thread. Alternatively we could consider caching `ListWalletDir` result and invalidate it when wallets are created or loaded.

yup, ugly quick fix and background thread seams reasonable.. but fix proves the point and lets be honest why should we waste time and risk locks, when traverse trough known node dir?.. .  also other GUI freezes should be tackled.
Before moon scale there is a long list of things in my backlog, that need to be fixed before bam..
.
There might be soon real users who use the gui,  and lets be real, gui is bitcoin for them.

promag commented on src/wallet/walletutil.cpp:
```diff
@@ -60,33 +60,69 @@ std::vector<fs::path> ListWalletDir()
+    fs::path sub_path = GetDataDir();
+
+    bool we_have_wallets_dir = (wallet_dir != sub_path);
+    const fs::path fblocks = GetBlocksDir();
```
Instead of multiple variables, you can declare something like `const std::vector ignore_paths{ GetBlocksDir(), sub_path / "testnet3", ... }` and then compare against this.

Saibato replied:
yup  :+1:  , btw strange how long those bare strings survived refactor, still no global #define BLOCKS_DIR or TESTNET_STR etc.

Sjors commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
-        const fs::path path = it->path().string().substr(offset);
+        // We dont want to iterate trough those special node dirs
+        if (it->status().type() == fs::directory_file) {
+            skip_path = false;
```
better to define `skip_path` here

Note to other reviewers, the enum file_type: https://www.boost.org/doc/libs/1_66_0/libs/filesystem/doc/reference.html#Enum-file_type

Sjors commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
+            skip_path = false;
+            for(const auto& fpath: ignore_paths) {
+                if (!skip_path && it->path().string().find(fpath.string()) == 0) {
+                    skip_path = true;
```
could use `break;` so you don't need to check `!skip_path &&` (and in fact, don't need it at all)

You can also do `it.disable_recursion_pending();` here

Saibato replied:
> (and in fact, don't need it at all)

hmm... If we allow wallets in those dir yes, but i would choose to touch the dir's not at all, 
Please view fixup I will post before i squash.
I guess we will need the skip  But sure about that we need some vote. :+1:

Sjors commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
+                it.disable_recursion_pending();
+                continue;
+            }
+        } else it.disable_recursion_pending();
```
nit: just use brackets and newline

Also, I think we should follow symlinks? `symlink_file`

luke-jr replied:
Following symlinks could create an infinite loop.

Sjors reviewed (commented):
Concept ACK

This seems orthogonal to using a background thread. We shouldn't be looking for wallets inside the blocks and chainstate dir even in the background.

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
+        if (it->status().type() == fs::directory_file) {
+            skip_path = false;
+            for(const auto& fpath: ignore_paths) {
+                if (!skip_path && it->path().string().find(fpath.string()) == 0) {
```
Since we're using exact paths in `ignore_paths`, can't we use an exact match? (And then `ignore_paths` can become an `unordered_set` and just use `.count` here...)

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
+    // Here we place the top level dirs we want to skip in case walletdir is datadir or blocksdir
+    const std::vector<fs::path> ignore_paths = {
+                                        blocks_dir,
+                                        data_dir / "testnet3",
+                                        data_dir / "regtest",
+                                        data_dir / "chainstate",
+                                        data_dir / "database",
+                                        data_dir / "indexes",
```
Suggest alphabetising.

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
+    // Here we place the top level dirs we want to skip in case walletdir is datadir or blocksdir
+    const std::vector<fs::path> ignore_paths = {
+                                        blocks_dir,
+                                        data_dir / "testnet3",
+                                        data_dir / "regtest",
+                                        data_dir / "chainstate",
+                                        data_dir / "database",
+                                        data_dir / "indexes",
```
Consider adding "blktree" and "coins" (see `files.md`)

luke-jr replied:
Maybe also the default blocks dir too

Saibato replied:
const std::vector<fs::path> ignore_paths = {
                                        blocks_dir,
                                        data_dir / "blktree",

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
             // Found a directory which contains wallet.dat btree file, add it as a wallet.
-            paths.emplace_back(path);
+            // Get wallet path relative to walletdir by removing walletdir from the wallet path.
+            // This can be replaced by boost::filesystem::lexically_relative once boost is bumped to 1.60.
```
Or hopefully `std::filesystem::path::lexically_relative` (C++17) :)

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,23 +57,48 @@ static bool IsBerkeleyBtree(const fs::path& path)
-        const fs::path path = it->path().string().substr(offset);
+        // We dont want to iterate trough those special node dirs
+        if (it->status().type() == fs::directory_file) {
+            skip_path = false;
+            for(const auto& fpath: ignore_paths) {
+                if (!skip_path && it->path().string().find(fpath.string()) == 0) {
+                    skip_path = true;
+                }
+            }
+            if (skip_path) {
+                it.disable_recursion_pending();
+                continue;
+            }
```
```suggestion
            if (ignore_paths.count(it->path()) {
                it.disable_recursion_pending();
                continue;
            }
```

Saibato replied:
:+1: Though I choose  
if (std::count(ignore_paths.begin(), ignore_paths.end(), it->path())) {
not sure ignore_paths has a count?

luke-jr replied:
It will once changed to an `unordered_set` as suggested

meshcollider commented:
Concept ACK, looks good to me

achow101 commented:
Concept ACK, please squash your commits.

Saibato commented:
squashed and rebased https://github.com/bitcoin/bitcoin/pull/19419/commits/caa418440dd5e996c730def08e807ddb5d504a44

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,59 @@ static bool IsBerkeleyBtree(const fs::path& path)
-                paths.emplace_back(path);
+        // We dont want to iterate trough those special node dirs
+        if (std::count(ignore_paths.begin(), ignore_paths.end(), it->path())) {
+            it.disable_recursion_pending();
```
`disable_recursion_pending` is a new name introduced in Boost 1.72. We support much older. Renaming this to `no_push` (the old name) should just work.

Saibato replied:
How about?

Suggestion:
``` Diff
diff --git a/src/wallet/walletutil.cpp b/src/wallet/walletutil.cpp
index a1387111e..55b81ed77 100644
--- a/src/wallet/walletutil.cpp
+++ b/src/wallet/walletutil.cpp
@@ -85,7 +85,13 @@ std::vector<fs::path> ListWalletDir()
 
         // We dont want to iterate trough those special node dirs
         if (std::count(ignore_paths.begin(), ignore_paths.end(), it->path())) {
-            it.disable_recursion_pending();
+            #ifdef USING_BOOST
+            // BOOST itterator
+                    it.no_push();
+            #else
+            // We have >= c++17
+                    it.disable_recursion_pending();
+            #endif
             continue;
         }
 

```

hebasto commented:
Concept ACK.

hebasto commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,59 @@ static bool IsBerkeleyBtree(const fs::path& path)
+                                        data_dir / "coins",
+                                        data_dir / "database",
+                                        data_dir / "indexes",
+                                        data_dir / "regtest",
+                                        data_dir / "testnet3"
```
Please add "signet".

hebasto commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,59 @@ static bool IsBerkeleyBtree(const fs::path& path)
     boost::system::error_code ec;
 
+    // Here we place the top level dirs we want to skip in case walletdir is datadir or blocksdir
+    // Those directory's are referenced in doc/files.md
```
nit:
```suggestion
    // Those directories are referenced in doc/files.md
```

hebasto commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,59 @@ static bool IsBerkeleyBtree(const fs::path& path)
-                // opened in a shared database environment for backwards compatibility.
-                // Add it to the list of available wallets.
-                paths.emplace_back(path);
+        // We dont want to iterate trough those special node dirs
```
nit:
```suggestion
        // We don't want to iterate through those special node dirs
```

hebasto commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,59 @@ static bool IsBerkeleyBtree(const fs::path& path)
+                    paths.emplace_back(it->path().filename());
+                }
             }
+        } catch (const std::exception& e) {
```
I personally prefer "throwing" versions of the Boost.Filesystem API. But I don't think that combining "throwing" and "non-throwing" versions of the Boost.Filesystem API in the same function is a good design.

Saibato replied:
with catch boost::exception
and  dynamic_cast<std.exception i run with a recursive path in an non catched     
 EXCEPTION: N5boost10filesystem16filesystem_errorE
so unless u have a hint, i leave it that way, for now?

Saibato commented:
Rebased https://github.com/bitcoin/bitcoin/pull/19419/commits/da6eb3c2f8ed5c8e2a5acb878a1abcb9d6f1c02b and addressed reviews.

hebasto commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,60 @@ bool IsBerkeleyBtree(const fs::path& path)
-                // Add it to the list of available wallets.
-                paths.emplace_back(path);
+        // We don't want to iterate through those special node dirs
+        if (std::count(ignore_paths.begin(), ignore_paths.end(), it->path())) {
```
nit: [std::find](https://en.cppreference.com/w/cpp/algorithm/find) could return earlier.

Saibato replied:
yup. no insist here from my side, but nit too if "not"  *const time* could that sidechannel reveal where the wallet is or else?

Btw> since u review here and in  #18095, 18095 does fix some problems, but not the wallet of death, could we transfer attention off reviewers to this here and fix some bugs fast in one sweep, before adding there new ones?
I must admit my PR headlines lock often a bit harmless and just like minor nits, but when I way in and insist, the hut is burning. just saying .

hebasto replied:
I don't think that `std::count` _guaranties_ const time to prevent sidechannel reveal where the wallet is or else. But I'm not an expert in this field. Feel free to mark this comment as resolved.

Saibato replied:
Ok, then i resolve this here and fix the header. thx for review.

hebasto reviewed (changes requested):
Approach ACK da6eb3c2f8ed5c8e2a5acb878a1abcb9d6f1c02b.

Mind adding `#include <algorithm>` in separate header group?

Saibato committed:
add #include <algorithm> to src/wallet/walletutil.h

Designed to be used on ranges of elements.

Saibato commented:
Add `#include <algorithm>` to header.

Tyi @hebasto while dev this i encountered and ugly wallet and filesystem edge case. that is only partly fixed by #18095 but with fewer changes and final here and in #19502 u might reach out to @luke-jr or @laanwj for more info.

luke-jr commented on src/wallet/walletutil.cpp:
```diff
@@ -57,35 +57,60 @@ bool IsBerkeleyBtree(const fs::path& path)
+                    // software will never create these files but will allow them to be
+                    // opened in a shared database environment for backwards compatibility.
+                    // Add it to the list of available wallets.
+                    paths.emplace_back(it->path().filename());
```
Why did you change this?

Saibato replied:
its not changed just the correct indent.

Saibato replied:
If u mean the whole logic itself, its because there where an other wallet access issue not mentioned here fixed also, could be no longer valid since we had some grave changes in master, i have to check, but since the PR its closed yet, it makes sense to let all as is and open two new and or wait until its reopened and then help out with the review.

luke-jr commented on src/wallet/walletutil.h:
```diff
@@ -5,6 +5,7 @@
 #ifndef BITCOIN_WALLET_WALLETUTIL_H
 #define BITCOIN_WALLET_WALLETUTIL_H
 
+#include <algorithm>
```
Why did you add this here? If it's needed in the .cpp file, put it there...

Saibato replied:
;was keen to see some compiler warnings, i though that was the hint from @hebasto  to add `<algorithm>` addressed and changed back to .cpp that other bug we can fix later.

luke-jr commented:
Please make these changes:

* https://github.com/bitcoin/bitcoin/commit/5a8f19eb331acbb1f7e2f52fe786ed21b30297b6
* https://github.com/bitcoin/bitcoin/commit/4fb4fb716cd222ca8ae988d039328a66eb6e1f3b

laanwj commented:
Marking this as up for grabs. The concept to not iterate needlessly into directories that *should* not contain wallets definitely makes sense, but I think this makes unnecessary changes that are not well motivated.

Saibato commented:
Rebased and addressed https://github.com/bitcoin/bitcoin/commit/5a8f19eb331acbb1f7e2f52fe786ed21b30297b6  https://github.com/bitcoin/bitcoin/commit/4fb4fb716cd222ca8ae988d039328a66eb6e1f3b an comments, changed by this a lot so pls review again  

@luke-jr could not use ur code by merge and lost by fixup  ur commit entrys, i hope u don;t mind.

Saibato commented:
Rebased and addressed https://github.com/bitcoin/bitcoin/commit/5a8f19eb331acbb1f7e2f52fe786ed21b30297b6  https://github.com/bitcoin/bitcoin/commit/4fb4fb716cd222ca8ae988d039328a66eb6e1f3b an comments, changed by this a lot so pls review again  

@luke-jr could not use ur code by merge and lost by fixup  ur commit entrys, i hope u don;t mind.

Saibato commented:
> to find wallets, that consumes time and could cause an unresponsive node.

@laanwj "... cause an unresponsive node" in commit message was a hint, this PR is not about to fix something and must not be merged anyway but to point to some lines of code and its implications to gave others a chance to have a look in this and guess by themselfs.
So my question is now, will have we have a fix and sure its not done if we just skip the boost traversing?

in #19502 and #18095 we see some progress, but will that be enough and why have I to write this even, if it is to u so obvious?

luke-jr commented:
@Saibato GitHub won't let *anyone* reopen this since your branch changed. Can you `git push ??? 3f9cc0cd736e79c563dee3573f0dab01dac10622:wallet_351 -f` temporarily (you can re-push your new version after it's reopened; put your repo name in place of `???` of course).

Don't forget to rebase!

Also: You need to include <set> instead of <algorithm> ;)

luke-jr commented:
Here's a rebase for later: https://github.com/bitcoin/bitcoin/compare/master...luke-jr:listwalletdir_skip_data

Saibato commented:
@luke-jr tyi was the last week just counting my **sats** and had not  looked in this, thx. for the reopen hint.
btw look now why some wallets are missing when using the 0.21 rc1 tag, maybe that was the reason i changed also the scan for wallets logic here , but just forgot to mention that sggh...
